pybamm-team/PyBaMM
# Test Optional Dependencies

# Description
This PR adds a test module for optional dependencies just for the core version of PyBaMM.
Important changes made:
 - Remove anytree from `test_import_optional_dependency` (it's a required dependency, not an optional one).
 - Make `pandas` a required dependency. It is actually installed along with `xarray` (currently a required dependency).
   To verify it:
   ```bash
    python3.11 -m venv venv # create new environment
    source venv/bin/activate # access it
    pip install xarray # install xarray
    pip list # show installed packages
   ```
    This step avoids finding it among the optional dependencies when we parse them using `importlib.metadata('pybamm')`. Since the core version of `pybamm` also installs it, the test to check that it does not contain optional dependencies would fail.
 - Add a separate test module. The reason for this is that the test `TestDependencies.test_optional_dependencies` as implemented, does not use the function `import_optional_dependency`* and conceptually does not verify that a unit function is working correctly. It is checking that in the core version of `pybamm` we are not installing more dependencies than we expect. Moreover, it does not have to be inside a module with filename `test_*` because it would always be executed in any test session. Please, tell me what you think about this and if you suggest any alternatives.
 - Add a separate job to the `test_on_push.yml` workflow that launch it with just the core `pybamm` installation.

*If I'm not wrong, we cannot use `import_optional_dependency` because distribution packages extracted from `importlib.metadata` do not necessarily have a 1:1 mapping with import packages, as per [documentation](https://docs.python.org/3/library/importlib.metadata.html#package-distributions):
> Important: These are not necessarily equivalent to or correspond 1:1 with the top-level import package names that can be imported inside Python code. One distribution package can contain multiple import packages (and single modules), and one top-level import package may map to multiple distribution packages if it is a namespace package. You can use [package_distributions()](https://docs.python.org/3/library/importlib.metadata.html#package-distributions) to get a mapping between them.

Fixes #3865

## Type of change

- [ ] Optimization (back-end change that speeds up the code)

# Key checklist:

- [X] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`)
- [X] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [X] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)
